plugins/blockvolume/hostvol/hostvol_manager.go

```diff
@@ -143,9 +181,156 @@ func (g *GlusterVolManager) GetOrCreateHostingVolume(name string, minSizeLimit u
 		return nil, errors.New("volume has not been started")
 	}
 
+	key := volume.BlockPrefix + blkName
+	val := strconv.FormatUint(minSizeLimit, 10)
+	volInfo.Metadata[key] = val
+
+	resizeFunc := func(blockHostingAvailableSize, blockSize uint64) uint64 { return blockHostingAvailableSize - blockSize }
+	if err = UpdateBlockHostingVolumeSize(volInfo, minSizeLimit, resizeFunc); err != nil {
+		log.WithError(err).Error("failed in updating hostvolume _block-hosting-available-size metadata")
+		return nil, err
+	}
+
+	// Note that any further error exit conditions should undo the above hostsize change
 	if err := volume.AddOrUpdateVolume(volInfo); err != nil {
 		log.WithError(err).Error("failed in updating volume info to store")
 	}
 
 	return volInfo, nil
 }
+
+// DeleteBlockInfoFromBHV resets the available space on the bhv and also deletes the block entry in the metadata of the bhv
+// In this function, if the bhv is empty i.e. there are no blocks, then the bhv delete is initiated
+func (g *GlusterVolManager) DeleteBlockInfoFromBHV(hostVol string, blkName string, size uint64) error {
+	var (
+		clusterLocks = transaction.Locks{}
+		prune        = false
+	)
+
+	if err := clusterLocks.Lock(hostVol); err != nil {
+		log.WithError(err).Error("error in acquiring cluster lock")
+		return err
+	}
+	volInfo, err := volume.GetVolume(hostVol)
+	if err != nil {
+		log.WithError(err).Errorf("failed to get host volume info %s", hostVol)
+		clusterLocks.UnLock(context.Background())
+		return err
+	}
+
+	for k := range volInfo.Metadata {
+		if k == (volume.BlockPrefix + blkName) {
+			delete(volInfo.Metadata, k)
+		}
+	}
+
+	resizeFunc := func(blockHostingAvailableSize, blockSize uint64) uint64 { return blockHostingAvailableSize + blockSize }
+	if err = UpdateBlockHostingVolumeSize(volInfo, size, resizeFunc); err != nil {
+		log.WithFields(log.Fields{
+			"error": err,
+			"size":  size,
+		}).Error("error in resizing the block hosting volume")
+	}
+
+	// TODO: Also make sure volInfo.Metadata[volume.BlockPrefix*] has no keys left
+	availableSizeInBytes, err := strconv.ParseUint(volInfo.Metadata[volume.BlockHostingAvailableSize], 10, 64)
+	if err != nil {
+		clusterLocks.UnLock(context.Background())
+		return err
+	}
+	if availableSizeInBytes == volInfo.Capacity {
+		if g.hostVolOpts.AutoDelete {
+			volInfo.Metadata[volume.BlockHostMarkedForPrune] = "true"
+			prune = true
+		}
+	}
+
+	if err := volume.AddOrUpdateVolume(volInfo); err != nil {
+		log.WithError(err).Error("failed in updating volume info to store")
+		clusterLocks.UnLock(context.Background())
+		return err
+	}
+	clusterLocks.UnLock(context.Background())
+
+	if prune == true {
+		_ = g.pruneBHV(volInfo.Name, blkName, size)
```

Comment:
if we are not logging the error in think we don't need to return an error from pruneBHV